### ISSUE TITLE
[AMD][kernel][Tier 5] Port sol-execbench L2/051_seqlen-finetuned-reconstructed_hyena_complete_forward_block to FlyDSL

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/FlyDSL#219]
# FlyDSL implementation: port sol-execbench's `051_seqlen-finetuned-reconstructed_hyena_complete_forward_block` problem to FlyDSL

## Goal

Port the `051_seqlen-finetuned-reconstructed_hyena_complete_forward_block` problem from `nvidia/SOL-ExecBench` (subset `L2`) to FlyDSL as part of the sol-execbench → FlyDSL coverage sweep. This is one of 235 problems across 4 subsets (L1, L2, Quant, FlashInfer-Bench) being tracked for FlyDSL kernel coverage.

**Source:** `magicslabnu/Hyena-small-32k` (per `definition.json` `hf_id`)

**Description:** Complete end-to-end HyenaBlock forward pass that fuses all operations from input to output: residual management, layer normalizations with dtype conversions, the full Hyena operator (input projection, short conv, filter generation, FFT convolution, gating), MLP with GELU activation, and final residual addition.

## Hardware / workload

```text
Hardware:  1 × MI300X (gfx942) + 1 × MI355X (gfx950)
Models:    (none — kernel benchmark, not LLM workload)
Subset:    L2
Problem:   051_seqlen-finetuned-reconstructed_hyena_complete_forward_block
Source:    magicslabnu/Hyena-small-32k
```

**Const axes:**
  - `d_model` = 256
  - `d_inner` = 1024
  - `order` = 2
  - `l_max` = 32768
  - `short_filter_order` = 3
  - `filter_order` = 64
  - `emb_dim` = 5

**Inputs:**
  - `hidden_states` [batch_size, seq_len, d_model] float32 — Input hidden states
  - `norm1_weight` [d_model] float32 — LayerNorm 1 weight
  - `norm1_bias` [d_model] float32 — LayerNorm 1 bias
  - `norm2_weight` [d_model] float32 — LayerNorm 2 weight
  - `norm2_bias` [d_model] float32 — LayerNorm 2 bias
  - `in_proj_weight` [inner_width, d_model] float32 — Input projection weight
  - `in_proj_bias` [inner_width] float32 — Input projection bias
  - `short_conv_weight` [inner_width, 1, short_filter_order] float32 — Short conv weight
  - `short_conv_bias` [inner_width] float32 — Short conv bias
  - `filter_linear1_weight` [filter_order, emb_dim] float32 — Filter MLP layer 1 weight
  - `filter_linear1_bias` [filter_order] float32 — Filter MLP layer 1 bias
  - `sin_freq` [1, filter_order] float32 — Sinusoidal activation frequency
  - `filter_linear2_weight` [filter_order, filter_order] float32 — Filter MLP layer 2 weight
  - `filter_linear2_bias` [filter_order] float32 — Filter MLP layer 2 bias
  - `filter_linear3_weight` [filter_order, filter_order] float32 — Filter MLP layer 3 weight
  - `filter_linear3_bias` [filter_order] float32 — Filter MLP layer 3 bias
  - `filter_linear_final_weight` [d_model, filter_order] float32 — Filter final projection weight
  - `filter_bias` [d_model] float32 — Filter bias for FFT convolution
  - `exp_mod_deltas` [1, 1, d_model] float32 — Exponential modulation deltas
  - `out_proj_weight` [d_model, d_model] float32 — Output projection weight
  - `out_proj_bias` [d_model] float32 — Output projection bias
  - `mlp_fc1_weight` [d_inner, d_model] float32 — MLP fc1 weight
  - `mlp_fc1_bias` [d_inner] float32 — MLP fc1 bias
  - `mlp_fc2_weight` [d_model, d_inner] float32 — MLP fc2 weight
  - `mlp_fc2_bias` [d_model] float32 — MLP fc2 bias
  - `layer_norm_eps` [] float32 — LayerNorm epsilon
  - `exp_mod_shift` [] float32 — Exponential modulation shift

**Outputs:**
  - `output` [batch_size, seq_len, d_model] float32 — Output hidden states

**Workload samples (first 3):**
  - {"uuid": "ecdc9dd6-141d-55b6-b702-ed60c6728e2f", "axes": {"batch_size": 1, "seq_len": 1024}, "inputs": {"hidden_states": {"type": "custom"}, "norm1_weight": {"type": "custom"}, "norm1_bias": {"type": 
  - {"uuid": "63675743-a638-56b7-90d0-f89c8a382b8e", "axes": {"batch_size": 2, "seq_len": 4096}, "inputs": {"hidden_states": {"type": "custom"}, "norm1_weight": {"type": "custom"}, "norm1_bias": {"type": 
  - {"uuid": "94f6ab59-8b9e-5fc7-ba70-f3fd22d24aa1", "axes": {"batch_size": 16, "seq_len": 512}, "inputs": {"hidden_states": {"type": "custom"}, "norm1_weight": {"type": "custom"}, "norm1_bias": {"type": 

```text
Baseline:  sol-execbench PyTorch reference (reference.py) on AMD with torch.compile off
Target:    max_abs_diff < 1e-2 vs PyTorch reference on all tested shapes
           kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
```

## Scope

1. `kernels/sol_execbench/l2_051_seqlen-finetuned-reconstructed_hyena_complete_forward_block.py` — FlyDSL kernel implementing the op described in `definition.json`, matching the PyTorch reference's numerics
2. `tests/kernels/test_sol_execbench_l2_051_seqlen-finetuned-reconstructed_hyena_complete_forward_block.py` — parity harness vs PyTorch reference, testing multiple workload shapes
3. `tests/kernels/bench_sol_execbench_l2_051_seqlen-finetuned-reconstructed_hyena_complete_forward_block.py` — reports `kernel_ms` per shape + comparison vs PyTorch eager baseline
4. `assessment.md` (≤300 words) — ranking which shapes FlyDSL wins/ties/loses vs PyTorch eager

## Acceptance

- Correctness: max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes
- Performance: kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
- Coverage: works on both gfx942 (MI300X) and gfx950 (MI355X)

## Out of scope

- Backward pass (unless the problem is itself a backward op)
- Weight gradients
- Optimizer step
- Cross-arch comparison vs NVIDIA H100 (PyTorch eager on AMD is the baseline)

## Reproduction

Clone `amdpilot-org/SOL-ExecBench`, download the dataset (`scripts/download_data.sh`), find the problem at `data/benchmark/L2/051_seqlen-finetuned-reconstructed_hyena_complete_forward_block/`, run the sol-execbench driver against a FlyDSL solution. Requires the unified aiter+triton+flydsl Docker image at `/home/xiasun/unified-rocm.Dockerfile` (not yet built).

## Definition of done

- [ ] Kernel implemented at `kernels/sol_execbench/l2_051_seqlen-finetuned-reconstructed_hyena_complete_forward_block.py`
- [ ] Tests pass on MI300X (gfx942)
- [ ] Tests pass on MI355X (gfx950)
- [ ] Bench shows ≥80% win rate vs PyTorch eager
- [ ] `assessment.md` written

## ## Leaderboard baseline to beat

These are the current best public results on NVIDIA B200 (from the
[sol-execbench leaderboard](https://research.nvidia.com/benchmarks/sol-execbench/leaderboard))
for problem `051_seqlen-finetuned-reconstructed_hyena_complete_forward_block` (kernel_id=145). The agent's
FlyDSL kernel on MI300X/MI355X should aim to beat the **Scoring Baseline**
(optimized PyTorch) and approach the **top human/AI submissions**, normalized
by cross-arch roofline.

```text
GPU:           NVIDIA B200 (leaderboard) vs AMD MI300X/MI355X (this task)
Latency unit:  ms

SOL Bound (theoretical roofline):     0.002648 ms   (SOL-Score=1.0000, speedup=243.81×)
Scoring Baseline (optimized PyTorch): 0.412016 ms   (SOL-Score=0.5588, speedup=1.27×)
Reference Implementation (plain PyTorch): 1.044071 ms   (SOL-Score=0.3351, speedup=0.52×)

Top public submissions:
  - **Rank 1** (`Hyra`): 0.021479 ms, SOL-Score=0.9659, speedup=24.29× vs reference
  - **Rank 2** (`Proud Wren @ NVIDIA`): 0.101193 ms, SOL-Score=0.8405
  - **Rank 3** (`doubleAI`): 0.109777 ms, SOL-Score=0.8291
```

**Cross-arch note:** B200 timings are not directly comparable to MI300X/MI355X
(different GPU arch, different roofline). The agent should:
1. Measure the PyTorch reference on MI300X/MI355X (same shapes, same Docker image)
2. Compute the FlyDSL kernel's speedup vs the on-AMD PyTorch reference
3. Compare that speedup ratio to the leaderboard's `avg_speedup` ratios above
4. Target: FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × (Rank-1 speedup vs PyTorch-reference-on-B200)

## Definition of done

- [ ] FlyDSL kernel passes correctness (max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes)
- [ ] FlyDSL kernel_ms measured on MI300X (gfx942) AND MI355X (gfx950) across >=5 workload shapes
- [ ] FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × Rank-1 speedup vs PyTorch-reference-on-B200 (=24.29× if top1 exists)
- [ ] `assessment.md` (<=300 words) ranks FlyDSL vs PyTorch eager on each shape, identifies win/tie/lose regimes
- [ ] Patch modifies actual kernel source (not just bench scripts) — bench-only patches are REJECTED

## Non-goals / rejection criteria

- Bench-only patches (no FlyDSL kernel source changes) are rejected
- Patches that set expected_value = measured_value (tautological pass) are rejected
- The metric contract must compare FlyDSL vs PyTorch reference on AMD, not an absolute ms
- "Completed" without a measured speedup ratio vs the on-AMD PyTorch reference is rejected

**Comment by sunxxuns** [comment]
## AMDPilot acceptance target

**The deliverable is a FlyDSL kernel.** A PyTorch reimplementation of this
problem -- with or without `torch.compile` -- is the *baseline*, not a result.
It does not close this issue no matter how closely it matches the reference.

### Implementation contract (hard requirement)

The submitted kernel MUST be written in FlyDSL and go through the FlyDSL
compiler. Concretely, the kernel module has to actually use the DSL, e.g.:

```python
import flydsl.compiler as flyc
from flydsl.expr import arith as ea, gpu, vector as ev
```

and dispatch through `flyc.jit` / `flyc.kernel`. A file placed under
`FlyDSL/kernels/sol_execbench/` that only calls `torch.*` ops does NOT satisfy
this issue, even if it is numerically perfect and even if it is faster.

If you conclude the port is not expressible in FlyDSL today, say so explicitly
and report exactly which DSL capability is missing. That is a legitimate,
useful outcome. Silently substituting a PyTorch implementation is not.

### Baseline definition

The baseline is the sol-execbench PyTorch reference (`reference.py`) running
eager, with `torch.compile` OFF. Measure it yourself in this run, on the same
GPU and the same workload shapes you benchmark your kernel on. Do not carry a
number over from another run, another GPU, or the issue text.

### Performance target

Your FlyDSL kernel must be **at least 10% faster than that measured baseline**
-- `kernel_ms <= 0.90 x baseline_ms` -- on **at least 80% of the tested
workload shapes**.

### Correctness target

`max_abs_diff < 1e-2` versus the PyTorch reference on **all** tested shapes. A
faster kernel that fails this is a failure, not a tradeoff.

### What you must report

Report a per-shape table containing, for every tested shape: the measured
eager baseline ms, your FlyDSL kernel ms, the ratio, and win/loss. Then state
the win rate against the 80% bar. A result without the baseline column next to
the kernel column is not checkable and will not be credited.
